### PR TITLE
Add uniqTheta in performance test

### PR DESCRIPTION
**Pull request by Avogar**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Not for changelog (changelog entry is not required)

### CODE REVIEW

**Comment by alexey-milovidov** [comment]
Would you mind renaming function from `uniqThetaSketch` to `uniqSketch`?
Because most of the other `uniq` functions are also sketches but don't have "sketch" in their names.

**Comment by alexey-milovidov** [comment]
How comes 14 seconds for 100 million records not insanely slow?

**Comment by Avogar** [comment]
> Would you mind renaming function from `uniqThetaSketch` to `uniqTheta`?
> Because most of the other `uniq` functions are also sketches but don't have "sketch" in their names.

Ok, will rename when we fix issue with uniqThetaSketch.

**Comment by alexey-milovidov** [comment]
Some queries have to be adjusted...

File: tests/performance/uniq.xml [hunk 1]
```diff
@@ -23,7 +23,6 @@
                 -->
                <value>SearchEngineID</value>
                <value>RegionID</value>
-               <value>SearchPhrase</value>
```
Comment (alexey-milovidov) [inline tests/performance/uniq.xml:26]: It has removed SearchPhrase key from all aggregate functions (uniq, uniqExact, uniqCombined, etc).

**Comment by alexey-milovidov** [comment]
Perf test has shown that uniqTheta is too slow on some use cases. It's interesting to figure out, why.

**Comment by alexey-milovidov** [comment]
@Mergifyio update

**Comment by alexey-milovidov** [comment]
Now the queries are too slow.

**Comment by alexey-milovidov** [comment]
@Avogar Now the queries are too fast.

Let me highlight the issue:
- only one query using `uniqTheta` is significantly slower than other queries.

We need to figure out why.
If we cannot make it faster - only that one query has to be adjusted in perf test.

**Comment by alexey-milovidov** [comment]
@Avogar it still needs your changes...

**Comment by Avogar** [comment]
@Mergifyio update

**Comment by Avogar** [comment]
I tried to figure out why only one query with uniqTheta is slower than the other queries, but I couldn't find the reason. Maybe the author can look at it. @pingyu do you have an idea why? (look at the report https://clickhouse-test-reports.s3.yandex.net/23311/6b2b2b09c3efb14b3d2b49cde992476a8710b910/performance_comparison/all-queries.html)

**Comment by pingyu** [comment]
> I tried to figure out why only one query with uniqTheta is slower than the other queries, but I couldn't find the reason. Maybe the author can look at it. @pingyu do you have an idea why? (look at the report https://clickhouse-test-reports.s3.yandex.net/23311/6b2b2b09c3efb14b3d2b49cde992476a8710b910/performance_comparison/all-queries.html)

Let me see see.

**Comment by pingyu** [comment]
> I tried to figure out why only one query with uniqTheta is slower than the other queries, but I couldn't find the reason. Maybe the author can look at it. @pingyu do you have an idea why? (look at the report https://clickhouse-test-reports.s3.yandex.net/23311/6b2b2b09c3efb14b3d2b49cde992476a8710b910/performance_comparison/all-queries.html)

Can I get the dataset `hits_100m_single` for further analysis ? @Avogar 

I cannot reproduce the issue with "hits_v1" from [here](https://clickhouse.tech/docs/en/getting-started/example-datasets/metrica/).

With a 355m dataset produced by inserting "hits_v1" 20 times, I got the performance: `uniqTheta(UserID) group by SearchPhrase` is only about 40% slower than `group by RegionID`, or `uniqCombined(18)`.

Machine: x64 8 cores, 16GB memory, [Tencent Linux](https://intl.cloud.tencent.com/document/product/213/32485) server.
ClickHouse: standalone server, built from latest master (4b1b612cfd) by llvm11.
```bash
$ clickhouse-client --query "SELECT count(1) from datasets.hits_v1m"
354955920
$
$ time clickhouse-client --query "SELECT SearchPhrase AS k, uniqTheta(UserID) FROM datasets.hits_v1m GROUP BY k FORMAT Null"

real    0m1.731s
user    0m0.016s
sys     0m0.004s
$ time clickhouse-client --query "SELECT SearchPhrase AS k, uniqCombined(18)(UserID) FROM datasets.hits_v1m GROUP BY k FORMAT Null"

real    0m1.327s
user    0m0.020s
sys     0m0.000s
$ time clickhouse-client --query "SELECT RegionID AS k, uniqTheta(UserID) FROM datasets.hits_v1m GROUP BY k FORMAT Null"

real    0m1.204s
user    0m0.012s
sys     0m0.008s
```



At the same time, after read the performance report and browsed related codes, I got the following:

* First, I think the issue happened in `merge`, as elapse time during `insert` should be the same as `group by RegionID`.

* Second, NDV of `SearchPhrase` is much larger than `RegionID`, which should lead to much more `merge`.

* Third, the implementation of `union` in "theta sketch" requires [`get_result()`](https://github.com/ClickHouse/ClickHouse/blob/bf482f75a708a10fea8657f26ef5bc8756135b72/src/AggregateFunctions/ThetaSketchData.h#L81) of another `theta_union`, which needs allocating, copying and sorting on 2^12 elements (see [here](https://github.com/ClickHouse-Extras/datasketches-cpp/blob/master/theta/include/theta_union_base_impl.hpp#L61)), and is possibly lead to high overhead.

I will verify these speculations after reproduce the issue.

**Comment by Avogar** [comment]
@pingyu You can download prepared partitions of this table. See https://clickhouse.tech/docs/en/getting-started/example-datasets/metrica/

Link to download: https://datasets.clickhouse.tech/hits/partitions/hits_100m_obfuscated_v1.tar.xz.

**Comment by pingyu** [comment]
> @pingyu You can download prepared partitions of this table. See https://clickhouse.tech/docs/en/getting-started/example-datasets/metrica/
> 
> Link to download: https://datasets.clickhouse.tech/hits/partitions/hits_100m_obfuscated_v1.tar.xz.

Great, thanks!

**Review by alexey-milovidov — APPROVED** [review]

**Comment by alexey-milovidov** [comment]
Conclusion: Apache Datasketches library is not so well battle tested on various scenarios as ClickHouse.

**Comment by pingyu** [comment]
I'm still trying to find out the reason.

Now I believe it's caused by the implementation of union operation of Theta sketches. I'm writing a program to verify the cause, and will try to propose a solution to Datasketches library later.

Besides, [perf](https://github.com/torvalds/linux/tree/master/tools/perf) cannot unwind the full stack of ClickHouse. It took me difficult to do analysis right on the performance test case. I would write an issue providing more information to ask for help.

**Pull request merged**